### PR TITLE
Add XML support [encoder]

### PR DESCRIPTION
**Pull request by Baccata**
* Added an XmlDocument.Encoder construct
* Created an XmlEncoderSchemaVisitor
* Derive an instance of XmlDocument.Encoder from schema
* Fix a few things in the decoding logic

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -975,7 +975,7 @@ def genSmithyImpl(config: Configuration) = Def.task {
       .map(_.data)
 
   val mc = "smithy4s.codegen.cli.Main"
-  val s = streams.value
+  val s = (config / streams).value
```
Comment (Baccata) [inline build.sbt:978]: kudos to @kubukoz for the pointer

File: modules/xml/src/smithy4s/xml/XmlDocument.scala [hunk 2]
```diff
@@ -47,11 +52,11 @@ object XmlDocument {
...
   // format: off
-  sealed trait XmlContent
-  case class XmlText(text: String)                                                       extends XmlContent
-  case class XmlElem(name: XmlQName, attributes: List[XmlAttr], children: List[XmlContent]) extends XmlContent
-  case class XmlAttr(name: XmlQName, value: List[XmlText])
-  case class XmlQName(prefix: Option[String], name: String) {
+  sealed trait XmlContent extends Product with Serializable
+  final case class XmlText(text: String)                                                       extends XmlContent
+  final case class XmlElem(name: XmlQName, attributes: List[XmlAttr], children: List[XmlContent]) extends XmlContent
+  final case class XmlAttr(name: XmlQName, value: List[XmlText]) extends XmlContent
```
Comment (Baccata) [inline modules/xml/src/smithy4s/xml/XmlDocument.scala:58]: this really helps simplify the encoding logic, at the cost of being a little bit less correct, modelling wise

File: modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala [hunk 3]
```diff
@@ -481,43 +492,73 @@ object XmlDecodingSpec extends SimpleIOSuite {
...
+            XmlDecodeError(XPath.root, "Expected Foo XML root element, got Bar")
+          )
         )
-      )
-    }
+      }
   }
 
-  def checkContent[A: Schema](xml: String, expected: A)(implicit
+  def checkContent[A: Schema](xmlString: String, expected: A)(implicit
```
Comment (Baccata) [inline modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala:508]: Switched the logic to test both encoding and decoding at the same time. The tests were already containing nearly everything we needed for that purpose.

File: modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala [hunk 4]
```diff
@@ -16,75 +16,85 @@
...
+
+import scala.reflect.ClassTag
 
-object XmlDecodingSpec extends SimpleIOSuite {
+object XmlCodecSpec extends SimpleIOSuite {
+
+  implicit class SchemaOps[A](schema: Schema[A]) {
+    def named(name: String) = schema.withId(ShapeId("default", name))
+    def x = named("x")
+    def n(implicit ct: ClassTag[A]) = named(
```
Comment (Baccata) [inline modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala:43]: Added this to derive a shapeId from the classtag of the type the Schema works against

File: modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala [hunk 5]
```diff
@@ -481,43 +494,73 @@ object XmlDecodingSpec extends SimpleIOSuite {
...
+
+      import cats.Show
+      implicit val showXmlDocument: Show[XmlDocument] = new Show[XmlDocument] {
+        def show(xmlDocument: XmlDocument): String =
+          XmlDocument.documentEventifier
+            .eventify(xmlDocument)
+            .compile
+            .toVector
+            .map(_.show)
+            .mkString("")
```
Comment (ybasket) [inline modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala:525]:
```suggestion
            .through(render())
            .compile
            .string
```

Has the advantage of using (you can opt-out) self-closing tags for more concise output. Added recently in fs2-data 1.5.1.

File: modules/xml/src/smithy4s/xml/XmlDocument.scala (hunk 2 above)
Comment (daddykotex) [reply modules/xml/src/smithy4s/xml/XmlDocument.scala:59]: minor: `values` ?

File: modules/xml/src/smithy4s/xml/XmlDocument.scala [hunk 6]
```diff
@@ -76,6 +82,14 @@ object XmlDocument {
     }
   }
 
+  private def getRootName[A](schema: Schema[A]): XmlQName = {
+    schema.hints
+      .get(XmlName)
+      .map(_.value)
+      .map(XmlQName.parse)
```
Comment (kubukoz) [inline modules/xml/src/smithy4s/xml/XmlDocument.scala:89]: totally nitpicky suggestion: extract these to a val so it's clearer that `getOrElse` is a fallback for the whole thing

File: modules/xml/src/smithy4s/xml/XmlDocument.scala [hunk 7]
```diff
@@ -111,6 +120,41 @@ object XmlDocument {
...
 
+  trait Encoder[A] {
+    def encode(value: A): XmlDocument
+  }
+  object Encoder {
+    def fromSchema[A](schema: Schema[A]): Encoder[A] = {
+      val rootName: XmlQName = getRootName(schema)
+      val xmlEncoder = XmlEncoderSchemaVisitor(schema)
+      if (xmlEncoder.encodesUnion) {
+        // The union encoder should always return a single XML element
```
Comment (kubukoz) [inline modules/xml/src/smithy4s/xml/XmlDocument.scala:131]: can we do a `require` for that?

File: modules/xml/src/smithy4s/xml/internals/XmlDecoder.scala [hunk 8]
```diff
@@ -39,11 +40,21 @@ import cats.data.NonEmptyList
 private[smithy4s] trait XmlDecoder[A] { self =>
   def decode(cursor: XmlCursor): Either[XmlDecodeError, A]
   final def map[B](f: A => B): XmlDecoder[B] = new XmlDecoder[B] {
+    override def down(name: XmlQName): XmlDecoder[B] = self.down(name).map(f)
```
Comment (kubukoz) [inline modules/xml/src/smithy4s/xml/internals/XmlDecoder.scala:43]: is this some sort of optimization?

File: modules/xml/src/smithy4s/xml/internals/XmlEncoderSchemaVisitor.scala [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+import smithy4s.schema._
+import smithy4s.{Schema => _, _}
+
+import XmlDocument._
+import cats.kernel.Monoid
+
+private[smithy4s] object XmlEncoderSchemaVisitor extends XmlEncoderSchemaVisitor
+
+private[smithy4s] abstract class XmlEncoderSchemaVisitor
+    extends SchemaVisitor[XmlEncoder]
```
Comment (kubukoz) [inline modules/xml/src/smithy4s/xml/internals/XmlEncoderSchemaVisitor.scala:34]: wanna make this cached?

same question for decoders in the other PR

File: modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala (hunk 3 above)
Comment (kubukoz) [reply modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala:510]: can we label them somehow to make errors clearer? i.e.g label one as an assertion on encoding, the other on decoding

File: modules/xml/src/smithy4s/xml/internals/XmlDecoder.scala (hunk 8 above)
Comment (Baccata) [reply modules/xml/src/smithy4s/xml/internals/XmlDecoder.scala:43]: Those were artifacts of a solution to a problem that ended up being solved differently. Will remove.

File: modules/xml/src/smithy4s/xml/XmlDocument.scala (hunk 7 above)
Comment (Baccata) [reply modules/xml/src/smithy4s/xml/XmlDocument.scala:131]: You mean a throwing one ?

File: modules/xml/src/smithy4s/xml/internals/XmlEncoderSchemaVisitor.scala (hunk 9 above)
Comment (Baccata) [reply modules/xml/src/smithy4s/xml/internals/XmlEncoderSchemaVisitor.scala:34]: Will do in a later PR, the XML artifact won't be published until we have a protocol interpreter using it

File: modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala (hunk 3 above)
Comment (Baccata) [reply modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala:510]: Weaver doesn't have a combinator for this and I don't want to make a bespoke one. However we have `traced(here)` which will capture the accurate position of the expectation (source locations are kept in a list in weaver, so you can trace the calls reasonably accurately)

File: modules/xml/src/smithy4s/xml/internals/XmlDecoder.scala (hunk 8 above)
Comment (Baccata) [reply modules/xml/src/smithy4s/xml/internals/XmlDecoder.scala]: [d45899d](https://github.com/disneystreaming/smithy4s/pull/503/commits/d45899dc33844c1b601d57e2076470c2af4b1a94)

File: modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala (hunk 3 above)
Comment (Baccata) [reply modules/xml/test/src/smithy4s/xml/XmlCodecSpec.scala:510]: [d45899d](https://github.com/disneystreaming/smithy4s/pull/503/commits/d45899dc33844c1b601d57e2076470c2af4b1a94)

File: modules/xml/src/smithy4s/xml/XmlDocument.scala (hunk 2 above)
Comment (Baccata) [reply modules/xml/src/smithy4s/xml/XmlDocument.scala]: [d45899d](https://github.com/disneystreaming/smithy4s/pull/503/commits/d45899dc33844c1b601d57e2076470c2af4b1a94)

File: modules/xml/src/smithy4s/xml/XmlDocument.scala (hunk 7 above)
Comment (kubukoz) [reply modules/xml/src/smithy4s/xml/XmlDocument.scala:131]: yeah just for sanity

**Review by kubukoz — APPROVED** [review]

**Pull request merged**